ethcore/sync/src/snapshot_sync.rs

```diff
@@ -57,111 +73,117 @@ impl Snapshot {
 		}
 
 		trace!(
-			target: "snapshot",
-			"Snapshot is now initialized with {} completed chunks.",
-			self.completed_chunks.len(),
+			target: "snapshot_sync",
+			"Snapshot initialized. {}/{} completed chunks.",
+			self.completed_chunks.len(), total_chunks
 		);
-
+		self.total_chunks = Some(total_chunks);
 		self.initialized = true;
 	}
 
-	/// Clear everything.
+	/// Clear everything and set `initialized` to false.
 	pub fn clear(&mut self) {
 		self.pending_state_chunks.clear();
 		self.pending_block_chunks.clear();
 		self.downloading_chunks.clear();
 		self.completed_chunks.clear();
 		self.snapshot_hash = None;
+		self.total_chunks = None;
 		self.initialized = false;
 	}
 
-	/// Check if currently downloading a snapshot.
+	/// Check if we're currently downloading a snapshot.
 	pub fn have_manifest(&self) -> bool {
 		self.snapshot_hash.is_some()
 	}
 
-	/// Reset collection for a manifest RLP
+	/// Clear the `Snapshot` and reset it with data from a `ManifestData` (i.e. the lists of
+	/// block&state chunk hashes contained in the `ManifestData`).
 	pub fn reset_to(&mut self, manifest: &ManifestData, hash: &H256) {
 		self.clear();
-		self.pending_state_chunks = manifest.state_hashes.clone();
-		self.pending_block_chunks = manifest.block_hashes.clone();
+		self.pending_state_chunks = HashSet::from_iter(manifest.state_hashes.clone());
+		self.pending_block_chunks = HashSet::from_iter(manifest.block_hashes.clone());
+		self.total_chunks = Some(self.pending_block_chunks.len() + self.pending_state_chunks.len());
 		self.snapshot_hash = Some(hash.clone());
 	}
 
-	/// Validate chunk and mark it as downloaded
+	/// Check if the the chunk is known, i.e. downloaded already or currently downloading; if so add
+	/// it to the `completed_chunks` set.
+	/// Returns a `ChunkType` with the hash of the chunk.
 	pub fn validate_chunk(&mut self, chunk: &[u8]) -> Result<ChunkType, ()> {
 		let hash = keccak(chunk);
 		if self.completed_chunks.contains(&hash) {
-			trace!(target: "sync", "Ignored proccessed chunk: {:x}", hash);
-			return Err(());
+			trace!(target: "snapshot_sync", "Already proccessed chunk {:x}. Ignoring.", hash);
+			return Ok(ChunkType::Dupe(hash));
```

Comment:
I'm not sure about this change. It would mean that a peer can just continuously send the same chunk over and over, potentially eating our bandwidth, without getting disconnected. Is it actually happening in the wild that peers would send twice the same chunk and get disconnected from for it?

Reply:
> Is it actually happening in the wild that peers would send twice the same chunk and get disconnected from for it?

Indeed that is the question. I have never seen it happen and I have never heard of anyone that reported this. It seems like an odd thing to add a check for unless it actually happened.

Wrt to the dos potential here: any malicious node could send around fake manifests and spray the network with bad snapshots and users will not know the blocks are bad until they've imported them all. Protecting against the case of duplicate chunks seems like worrying about the wrong thing if you see what I mean?
On the other hand I also don't see how a non-malicious user could end up sending a duplicate unless there's a bug in our code.

(Somewhat related, on Ropsten there are snapshots at `#6500000` that are bad, most likely taken from a node on the wrong fork; it's quite frustrating to see your node go through the whole restoration and start syncing from `#6500001` and fail.)

I don't have a strong opinion here. @ordian what do you think? Should I undo this change?

Reply:
Fair enough, reverted.